Bad, Sad error with bitwise-and-expr when one operand is byte and the other is int

**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/io;

public function main(string... args) {
    int i = 2;
    byte j = 6;

    byte k = i & j;
    io:println(k);
}
```

ballerina-internal.log
```
[2019-08-19 22:48:07,579] ERROR {org.ballerinalang.jvm.scheduling.Scheduler} - Strand died 
java.lang.ArrayIndexOutOfBoundsException: -1
	at org.objectweb.asm.Frame.merge(Frame.java:1268)
	at org.objectweb.asm.Frame.merge(Frame.java:1244)
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1607)
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1543)
	at org.ballerinalang.nativeimpl.jvm.methodvisitor.VisitMaxs.visitMaxs(VisitMaxs.java:49)
	at ballerina.jvm.$value$MethodVisitor.visitMaxs(types.bal:61)
	at ballerina.jvm.$value$MethodVisitor.call(types.bal)
	at ballerina.compiler_backend_jvm.jvm_method_gen.genJMethodForBFunc(jvm_method_gen.bal:334)
	at ballerina.compiler_backend_jvm.jvm_method_gen.generateMethod(jvm_method_gen.bal:30)
	at ballerina.compiler_backend_jvm.jvm_package_gen.generatePackage(jvm_package_gen.bal:200)
	at ballerina.compiler_backend_jvm.main.generateJarBinary(main.bal:74)
	at ballerina.compiler_backend_jvm.main.main(main.bal:50)
	at ballerina.compiler_backend_jvm.___init.$lambda$main$(compiler_backend_jvm)
	at org.ballerinalang.jvm.scheduling.SchedulerItem.execute(Scheduler.java:452)
	at org.ballerinalang.jvm.scheduling.Scheduler.run(Scheduler.java:228)
	at org.ballerinalang.jvm.scheduling.Scheduler.runSafely(Scheduler.java:198)
	at java.lang.Thread.run(Thread.java:748)
 
[2019-08-19 22:48:07,604] ERROR {org.ballerinalang.tool.Main} - org.ballerinalang.jvm.util.RuntimeUtils.handleRuntimeThrowable(Ljava/lang/Throwable;)V 
java.lang.RuntimeException: org.ballerinalang.jvm.util.RuntimeUtils.handleRuntimeThrowable(Ljava/lang/Throwable;)V
	at org.ballerinalang.util.BootstrapRunner.generateJarBinaryViaCompiledBackend(BootstrapRunner.java:140)
	at org.ballerinalang.packerina.task.CreateJarTask.execute(CreateJarTask.java:65)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:298)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)
Caused by: java.lang.NoSuchMethodError: org.ballerinalang.jvm.util.RuntimeUtils.handleRuntimeThrowable(Ljava/lang/Throwable;)V
	at ballerina.compiler_backend_jvm.___init.main(compiler_backend_jvm)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.ballerinalang.util.BootstrapRunner.generateJarBinaryViaCompiledBackend(BootstrapRunner.java:137)
	... 5 more
```

**Affected Versions:**
1.0.0-beta-SNAPSHOT - 15th August build

--- a/compiler/ballerina-backend-jvm/src/main/ballerina/src/compiler_backend_jvm/jvm_instruction_gen.bal
+++ b/compiler/ballerina-backend-jvm/src/main/ballerina/src/compiler_backend_jvm/jvm_instruction_gen.bal
@@ -474,13 +474,20 @@ type InstructionGenerator object {
     }
 
     function generateBitwiseAndIns(bir:BinaryOp binaryIns) {
-        self.loadVar(binaryIns.rhsOp1.variableDcl);
-        self.loadVar(binaryIns.rhsOp2.variableDcl);
-
-        bir:BType opType = binaryIns.rhsOp1.typeValue;
-        if (opType is bir:BTypeInt) {
+        bir:BType opType1 = binaryIns.rhsOp1.typeValue;
+        bir:BType opType2 = binaryIns.rhsOp2.typeValue;
+        
+        if (opType1 is bir:BTypeInt && opType2 is bir:BTypeInt) {
+            self.loadVar(binaryIns.rhsOp1.variableDcl);
+            self.loadVar(binaryIns.rhsOp2.variableDcl);
             self.mv.visitInsn(LAND);
         } else {
+            self.loadVar(binaryIns.rhsOp1.variableDcl);
+            generateCheckCastToByte(self.mv, opType1);
+            
+            self.loadVar(binaryIns.rhsOp2.variableDcl);
+            generateCheckCastToByte(self.mv, opType2);
+            
             self.mv.visitInsn(IAND);
         }
         self.storeToVar(binaryIns.lhsOp.variableDcl);

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/binaryoperations/BinaryExprTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/binaryoperations/BinaryExprTest.java
@@ -146,9 +146,9 @@ public class BinaryExprTest {
         BValue[] args = {new BInteger(a), new BInteger(b), new BByte(c), new BByte(d)};
         BValue[] returns = BRunUtil.invoke(result, "bitwiseAnd", args);
 
-        Assert.assertEquals(((BInteger)returns[0]).intValue(), a & b);
-        Assert.assertEquals(((BByte)returns[1]).byteValue(), a & c);
-        Assert.assertEquals(((BByte)returns[2]).byteValue(), c & d);
-        Assert.assertEquals(((BByte)returns[3]).byteValue(), b & d);
+        Assert.assertEquals(((BInteger) returns[0]).intValue(), a & b);
+        Assert.assertEquals(((BByte) returns[1]).byteValue(), a & c);
+        Assert.assertEquals(((BByte) returns[2]).byteValue(), c & d);
+        Assert.assertEquals(((BByte) returns[3]).byteValue(), b & d);
     }
 }

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/binaryoperations/BinaryExprTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/binaryoperations/BinaryExprTest.java
@@ -17,6 +17,8 @@
 package org.ballerinalang.test.expressions.binaryoperations;
 
 import org.ballerinalang.model.values.BBoolean;
+import org.ballerinalang.model.values.BByte;
+import org.ballerinalang.model.values.BInteger;
 import org.ballerinalang.model.values.BValue;
 import org.ballerinalang.test.util.BCompileUtil;
 import org.ballerinalang.test.util.BRunUtil;
@@ -133,5 +135,20 @@ public class BinaryExprTest {
         boolean actualResult = ((BBoolean) returns[0]).booleanValue();
 
         Assert.assertEquals(actualResult, expectedResult);
+    }
+
+    @Test(description = "Test bitwise and")
+    public void bitwiseAndTest() {
+        int a = 5;
+        int b = 6;
+        byte c = 7;
+        byte d = 8;
+        BValue[] args = {new BInteger(a), new BInteger(b), new BByte(c), new BByte(d)};
+        BValue[] returns = BRunUtil.invoke(result, "bitwiseAnd", args);
+
+        Assert.assertEquals(((BInteger)returns[0]).intValue(), a & b);
+        Assert.assertEquals(((BByte)returns[1]).byteValue(), a & c);
+        Assert.assertEquals(((BByte)returns[2]).byteValue(), c & d);
+        Assert.assertEquals(((BByte)returns[3]).byteValue(), b & d);
     }
 }

--- a/tests/jballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/binary-expr.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/expressions/binaryoperations/binary-expr.bal
@@ -28,3 +28,12 @@ function multiBinaryORExpression(boolean one, boolean two, boolean three) return
 function multiBinaryExpression(boolean one, boolean two, boolean three) returns (boolean) {
     return (!one || (two && three)) || (!three || (one && two));
 }
+
+function bitwiseAnd(int a, int b, byte c, byte d) returns [int, byte, byte, byte] {
+    [int, byte, byte, byte] res = [0, 0, 0, 0];
+    res[0] = a & b;
+    res [1] = a & c;
+    res [2] = c & d;
+    res [3] = b & d;
+    return res;
+}

